feat: add CMSShape PDF

ikrommyd opened the pull request:
Starting a PR to add the CMSShape PDF into zfit-physics.
Related to https://github.com/zfit/zfit/issues/512


## Proposed Changes

  - Implement CMSShape

## Tests added

  - `tests/test_pdf_cmsshape.py`

## Checklist

 - [x] change approved
 - [x] implementation finished
 - [x] correct namespace imported
 - [x] tests added
 - [x] CHANGELOG updated

ikrommyd committed:
empty commit

ikrommyd committed:
first attempt

ikrommyd committed:
be more explicit

jonas-eschle commented on zfit_physics/models/pdf_cmsshape.py:
```diff
@@ -0,0 +1,68 @@
+from typing import Optional
+
+import numpy as np
+import tensorflow as tf
```
Hm, would be good to have znp. but erf and erfc isn't directly available in the namespace, we should add this in zfit. But TF is fine, it's straight forward to convert

ikrommyd replied:
Unless you want to add more stuff into the `znp` namespace we keep it that way for now.

jonas-eschle commented on zfit_physics/models/pdf_cmsshape.py:
```diff
@@ -0,0 +1,68 @@
+        name: str = "CMSShape",
+    ):
+        params = {"beta": beta, "gamma": gamma, "m": m}
+        super().__init__(obs=obs, params=params, name=name, extended=extended, norm=norm)
```
Looks good, what about the order of the parameters, i.e. `m`? We should follow as close as possible others, i.e. the voigt.

ikrommyd replied:
I thought about it. This the the original implementation: https://gitlab.cern.ch/cms-muonPOG/spark_tnp/-/blob/Spark3/RooCMSShape.cc. Here the `alpha` and `peak` are the redundant parameters thet are replaced by just `m`. `m` is a "loc" parameter and is the approcimate center of the distribution. `scipy` uses the name `loc` for such parameters, i.e. in Gaussian `pdf(x, loc=0, scale=1)` but in `zfit` I see `m` being used. We can put it first if you want and the naming convention is also up to debate.

jonas-eschle replied:
yes, tricky with the names, but as we have m in cauchy, voigt etc, that makes sense to keep it m here too. So m seems reasonable to me. Let's just put it first, so it's consisten with all the other m/mu

ikrommyd replied:
Done in latest commit.

jonas-eschle reviewed (commented):
Looks good overall, that's the right direction. Docs, (serialization), tests, but you know all that. And as discussed, feel free to add numba-stats as a dependency, that makes sense to test against

jonas-eschle commented on zfit_physics/models/pdf_cmsshape.py:
```diff
@@ -0,0 +1,68 @@
+
+
+def cmsshape_integral(limits: ztyping.SpaceType, params: dict, model) -> tf.Tensor:
+    lower, upper = limits.rect_limits
```
```suggestion
    lower, upper = limits.limit1d
```
I guess this is the culprit, it has to be of a lower dimension, this should work

ikrommyd replied:
It doees indeed fix the problem. But why doesn't it cause a problem here in `relbw`? https://github.com/zfit/zfit-physics/blob/251f4e2de542fcc753dc3ab29c3ae4e01a161815/zfit_physics/models/pdf_relbw.py#L134

ikrommyd committed:
fix `.pdf(x)` output shape

Co-authored-by: Jonas Eschle <jonas.eschle@cern.ch>

ikrommyd committed:
add some testing

ikrommyd committed:
more testing against numba-stats

ikrommyd committed:
add docstrings

ikrommyd commented:
@jonas-eschle I'm happy with the current state and testing. Please review it when you got some time.

jonas-eschle commented on requirements_dev.txt:
```diff
@@ -2,6 +2,7 @@ bumpversion>=0.5.3
 coverage>=4.5.1
 coverage>=4.5.1
 flake8>=3.5.0
+numba-stats @ git+https://github.com/HDembinski/numba-stats.git
```
```suggestion
numba-stats @ git+https://github.com/HDembinski/numba-stats.git  # CMSShape not yet released (expected 1.8.0)
```

jonas-eschle commented:
Looks good (minus minor things), and you reminded me to also increase the visibility of zfit-physics

ikrommyd committed:
Update requirements_dev.txt

Co-authored-by: Jonas Eschle <jonas.eschle@cern.ch>

ikrommyd committed:
put m parameter first

ikrommyd commented:
@jonas-eschle Alright, let the tests run and merge.

jonas-eschle committed:
remove typo

jonas-eschle committed:
docs: md to rst links

jonas-eschle reviewed (approved):
Looks good! I've just fixed two minor doc things: there was a piece of string laying around in the code and I've converted the docs from MD to RST, which is usually used in Python docstrings (afaik).

ikrommyd commented:
> Looks good! I've just fixed two minor doc things: there was a piece of string laying around in the code and I've converted the docs from MD to RST, which is usually used in Python docstrings (afaik).

Thanks, I saw md in the relativistic breit wigner implementation so I just followed that. Thanks for removing the string that was lying around, I had forgotten it. Merge it whenever you want!
